File: server/lib/gpt.py
import openai
from openai import OpenAI

client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")

# TODO: The 'openai.base_url' option isn't read in the client API. You will need to pass it when you instantiate the client, e.g. 'OpenAI(base_url='http://localhost:11434/v1')'
# openai.base_url = 'http://localhost:11434/v1'


def get_gpt_response(
    prompt, print_stream=False, print_func=None, model="mistral", temperature=0
):
    def _print_func(x):
        print_func(x) if print_func else print(x, end="")

    response = client.chat.completions.create(
        model=model,
        messages=[{"role": "user", "content": prompt}],
        temperature=temperature,
        stream=True if print_stream == True else False,
    )

    if print_stream:
        for chunk in response:
            content = chunk.choices[0].delta.content
            _print_func(content)
            whole_message += content

    return response.choices[0].message.content


def get_gpt_chat_response(
    messages, print_stream=False, print_func=None, model="mistral", temperature=0
):
    def _print_func(x):
        print_func(x) if print_func else print(x, end="")

    response = client.chat.completions.create(
        model=model,
        messages=messages,
        temperature=temperature,
        stream=True if print_stream == True else False,
    )

    whole_message = ""

    if print_stream:
        for chunk in response:
            if chunk.choices[0].delta.content is not None:
                content = chunk.choices[0].delta.content
                _print_func(content)
                whole_message += content

    return (
        whole_message if print_stream == True else response.choices[0].message.content
    )

File: src/lib/gpt.ts
export const getGptResponse = async (
  prompt: string,
  printStream = false,
  printFunc: ((text: string) => void) | null = null,
  model = "mistral",
  temperature = 0
): Promise<string> => {
  const response = await fetch("http://localhost:11434/v1/chat/completions", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model,
      messages: [{ role: "user", content: prompt }],
      temperature,
      stream: printStream,
    }),
  });

  if (printStream) {
    let wholeMessage = "";
    const reader = response.body.getReader();
    const decoder = new TextDecoder("utf-8");

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      const chunk = decoder.decode(value);
      const lines = chunk.split("\n");

      for (const line of lines) {
        if (line.startsWith("data:")) {
          const data = line.substring(5).trim();
          if (data === "[DONE]") {
            break;
          }
          try {
            const json = JSON.parse(data);
            const content = json.choices[0].delta.content;
            if (content) {
              if (printFunc) {
                printFunc(content);
              } else {
                process.stdout.write(content);
              }
              wholeMessage += content;
            }
          } catch (error) {
            console.error("Could not JSON parse stream data:", data, error);
          }
        }
      }
    }
    return wholeMessage;
  } else {
    const data = await response.json();
    return data.choices[0].message.content;
  }
};

export const getGptChatResponse = async (
  messages: { role: string; content: string }[],
  printStream = false,
  printFunc: ((text: string) => void) | null = null,
  model = "mistral",
  temperature = 0
): Promise<string> => {
  const response = await fetch("http://localhost:11434/v1/chat/completions", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model,
      messages,
      temperature,
      stream: printStream,
    }),
  });

  let wholeMessage = "";
  if (printStream) {
    const reader = response.body.getReader();
    const decoder = new TextDecoder("utf-8");

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      const chunk = decoder.decode(value);
      const lines = chunk.split("\n");

      for (const line of lines) {
        if (line.startsWith("data:")) {
          const data = line.substring(5).trim();
          if (data === "[DONE]") {
            break;
          }
          try {
            const json = JSON.parse(data);
            const content = json.choices[0].delta.content;
            if (content) {
              if (printFunc) {
                printFunc(content);
              } else {
                process.stdout.write(content);
              }
              wholeMessage += content;
            }
          } catch (error) {
            console.error("Could not JSON parse stream data:", data, error);
          }
        }
      }
    }
  } else {
    const data = await response.json();
    wholeMessage = data.choices[0].message.content;
  }
  return wholeMessage;
};
